src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala

```diff
@@ -0,0 +1,117 @@
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Success
+
+final case class EServicesApiServiceImpl(
+  agreementProcessService: AgreementProcessService,
+  catalogProcessService: CatalogProcessService,
+  tenantManagementService: TenantManagementService,
+  partyProcessService: PartyProcessService
+)(implicit ec: ExecutionContext)
+    extends EservicesApiService {
+
+  private implicit val logger: LoggerTakingImplicit[ContextFieldsToLog] =
+    Logger.takingImplicit[ContextFieldsToLog](this.getClass)
+
+  private val ACTIVE_DESCRIPTOR_STATES_FILTER: List[CatalogProcess.EServiceDescriptorState] = List(
+    CatalogProcess.EServiceDescriptorState.PUBLISHED,
+    CatalogProcess.EServiceDescriptorState.SUSPENDED,
+    CatalogProcess.EServiceDescriptorState.DEPRECATED
+  )
+
+  override def getEServicesCatalog(q: Option[String], producersIds: String, states: String, offset: Int, limit: Int)(
+    implicit
+    contexts: Seq[(String, String)],
+    toEntityMarshallerProblem: ToEntityMarshaller[Problem],
+    toEntityMarshallerCatalogEServices: ToEntityMarshaller[CatalogEServices]
+  ): Route = {
+    val result = for {
+      requesterId       <- getOrganizationIdFutureUUID(contexts)
+      apiStates         <- parseArrayParameters(states).traverse(EServiceDescriptorState.fromValue).toFuture
+      pagedResults      <- catalogProcessService.getEServices(
+        name = q,
+        producersIds = parseArrayParameters(producersIds),
+        states = apiStates.map(CatalogProcess.EServiceDescriptorState.fromApi),
+        offset = offset,
+        limit = limit
+      )
+      enhancedEServices <- Future.traverse(pagedResults.eservices)(enhanceEService(requesterId))
+    } yield CatalogEServices(
+      eservices = enhancedEServices,
+      pagination = Pagination(offset = offset, limit = limit, totalResults = pagedResults.totalCount)
+    )
+
+    onComplete(result) {
+      handleError(s"Error retrieving Catalog EServices") orElse { case Success(eServices) =>
+        getEServicesCatalog200(eServices)
+      }
+    }
+  }
+
+  private def enhanceEService(
+    requesterId: UUID
+  )(eService: CatalogProcess.EService)(implicit contexts: Seq[(String, String)]): Future[CatalogEService] = for {
+    // TODO Use directly the tenant once the name will be added to its model
+    producerTenant      <- tenantManagementService.getTenant(eService.producerId)
+    selfcareId          <- producerTenant.selfcareId.toFuture(MissingSelfcareId(producerTenant.id))
+    producerInstitution <- partyProcessService.getInstitution(selfcareId)
+    // End TODO
+
+    requesterTenant <-
+      if (requesterId != eService.producerId) tenantManagementService.getTenant(requesterId)
+      else Future.successful(producerTenant)
+
+    activeDescriptor = eService.descriptors
+      .filter(d => ACTIVE_DESCRIPTOR_STATES_FILTER.contains(d.state))
+      .sortBy(_.version.toInt)
+      .lastOption
+
+    agreement <- activeDescriptor.flatTraverse(d =>
```

Comment:
nope, that's sequential, it's better to use `Future.traverse(activeDescriptor)(d => ...).map(_.flatten)`